src/backtrace.rs

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use core::arch::asm;
+
+pub const MAX_BACKTRACE_ADDRESSES: usize = 10;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+pub(crate) fn backtrace() -> [Option<usize>; MAX_BACKTRACE_ADDRESSES] {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
+        tmp
+    };
+
+    backtrace_internal(fp)
+}
+
+fn backtrace_internal(fp: u64) -> [Option<usize>; MAX_BACKTRACE_ADDRESSES] {
+    let mut result = [None; MAX_BACKTRACE_ADDRESSES];
+    let mut index = 0;
+    let mut skip = 2;
+
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+
+    let mut fp = fp;
+    loop {
+        // Safe because we are only reading in the valid stack trace area
+        // Also, this is only called on panic
+
+        if fp < stack_start || fp > stack_end {
+            break;
+        }
+
+        let address = unsafe { (fp as *const u64).offset(-1).read_volatile() }; // RA/PC
+        fp = unsafe { (fp as *const u64).offset(-2).read_volatile() }; // next FP
+
+        if address == 0 {
+            break;
```

Comment:
Again, is this something that we could usefully report an error out on?

Reply:
again, enum

Reply:
fixed